File: examples/php/src/config_based_usage.php
<?php

declare(strict_types=1);

/**
 * Configuration-based usage example for Yosina PHP library.
 * This example demonstrates using direct transliterator configurations.
 */

require_once __DIR__ . '/../vendor/autoload.php';

use Yosina\Yosina;

function main(): void
{
    echo "=== Yosina PHP Configuration-based Usage Example ===\n\n";

    // Create transliterator with direct configurations
    $configs = [
        ["spaces", []],
        ["prolonged-sound-marks", ["replace_prolonged_marks_following_alnums" => true]],
        ["jisx0201-and-alike", []],
    ];

    $transliterator = Yosina::makeTransliterator($configs);

    echo "--- Configuration-based Transliteration ---\n";

    // Test cases demonstrating different transformations
    $testCases = [
        ["hello　world", "Space normalization"],
        ["カタカナーテスト", "Prolonged sound mark handling"],
        ["ＡＢＣ１２３", "Full-width conversion"],
        ["ﾊﾝｶｸ ｶﾀｶﾅ", "Half-width katakana conversion"],
    ];

    foreach ($testCases as [$testText, $description]) {
        $result = $transliterator($testText);
        echo "{$description}:\n";
        echo "  Input:  '{$testText}'\n";
        echo "  Output: '{$result}'\n";
        echo "\n";
    }

    // Demonstrate individual transliterators
    echo "--- Individual Transliterator Examples ---\n";

    // Spaces only
    $spacesOnly = Yosina::makeTransliterator([["spaces", []]]);
    $spaceTest = "hello　world　test";  // ideographic spaces
    echo "Spaces only: '{$spaceTest}' → '" . $spacesOnly($spaceTest) . "'\n";

    // Kanji old-new only
    $kanjiOnly = Yosina::makeTransliterator([
        ["ivs-svs-base", ["mode" => "ivs-or-svs", "charset" => "unijis_2004"]],
        ["kanji-old-new", []],
        ["ivs-svs-base", ["mode" => "base", "charset" => "unijis_2004"]],
    ]);
    $kanjiTest = "廣島檜";
    echo "Kanji only: '{$kanjiTest}' → '" . $kanjiOnly($kanjiTest) . "'\n";

    // JIS X 0201 conversion only
    $jisx0201Only = Yosina::makeTransliterator([
        ["jisx0201-and-alike", ['fullwidthToHalfwidth' => false]],
    ]);
    $jisxTest = "ﾊﾛｰ ﾜｰﾙﾄﾞ";
    echo "JIS X 0201 only: '{$jisxTest}' → '" . $jisx0201Only($jisxTest) . "'\n";
}

if (basename(__FILE__) === basename($_SERVER['SCRIPT_NAME'])) {
    main();
}

File: dart/example/config_based_usage.dart
// Configuration-based usage example for Yosina Dart library.
// This example demonstrates using direct transliterator configurations.

import 'package:yosina/yosina.dart';

void main() {
  print('=== Yosina Dart Configuration-based Usage Example ===\n');

  // Create transliterator with direct configurations (uses default registry)
  final transliterator = Transliterator.withMap([
    {'name': 'spaces'},
    {
      'name': 'prolongedSoundMarks',
      'options': {
        'replaceProlongedMarksFollowingAlnums': true,
      },
    },
    {'name': 'jisx0201AndAlike'},
  ]);

  print('--- Configuration-based Transliteration ---');

  // Test cases demonstrating different transformations
  final testCases = [
    ('hello　world', 'Space normalization'),
    ('カタカナーテスト', 'Prolonged sound mark handling'),
    ('ＡＢＣ１２３', 'Full-width conversion'),
    ('ﾊﾝｶｸ ｶﾀｶﾅ', 'Half-width katakana conversion'),
  ];

  for (final (testText, description) in testCases) {
    final result =
        Chars.charsToString(transliterator(Chars.fromString(testText)));
    print('$description:');
    print('  Input:  \'$testText\'');
    print('  Output: \'$result\'');
    print('');
  }

  // Demonstrate individual transliterators
  print('--- Individual Transliterator Examples ---');

  // Spaces only
  final spacesOnly = Transliterator.withList(['spaces']);
  const spaceTest = 'hello　world　test'; // ideographic spaces
  print(
      'Spaces only: \'$spaceTest\' → \'${Chars.charsToString(spacesOnly(Chars.fromString(spaceTest)))}\'');

  // Kanji old-new only
  final kanjiOnly = Transliterator.withMap([
    {
      'name': 'ivsSvsBase',
      'options': {
        'mode': 'ivs-svs',
        'charset': Charset.unijis2004,
      },
    },
    {'name': 'kanjiOldNew'},
    {
      'name': 'ivsSvsBase',
      'options': {
        'mode': 'base',
        'charset': Charset.unijis2004,
      },
    },
  ]);
  const kanjiTest = '廣島檜';
  print(
      'Kanji only: \'$kanjiTest\' → \'${Chars.charsToString(kanjiOnly(Chars.fromString(kanjiTest)))}\'');

  // JIS X 0201 conversion only
  final jisx0201Only = Transliterator.withMap([
    {
      'name': 'jisx0201AndAlike',
      'options': {
        'fullwidthToHalfwidth': false,
      },
    },
  ]);
  const jisxTest = 'ﾊﾛｰ ﾜｰﾙﾄﾞ';
  print(
      'JIS X 0201 only: \'$jisxTest\' → \'${Chars.charsToString(jisx0201Only(Chars.fromString(jisxTest)))}\'');
}
